[SPARK-42093][SQL] Move JavaTypeInference to AgnosticEncoders

hvanhovell opened the pull request:
### What changes were proposed in this pull request?
This PR makes `JavaTypeInference` produce an `AgnosticEncoder`. The expression generation for these encoders is moved to `ScalaReflection`.

### Why are the changes needed?
For the Spark Connect Scala Client we also want to be able to use Java Bean based results.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
I have added a lot of tests to `JavaTypeInferenceSuite`.

hvanhovell committed:
AgnosticEncoders

hvanhovell committed:
Merge remote-tracking branch 'apache/master' into SPARK-41690

hvanhovell committed:
Fix a couple of bugs

hvanhovell committed:
Boxed encoders.

hvanhovell committed:
Update doc, and add a proper field type.

hvanhovell committed:
WIP

hvanhovell committed:
Use Boxed values instead primitives in RowEncoder

hvanhovell committed:
Fix validation and harden array serialization

hvanhovell committed:
Remove println from code generator

hvanhovell committed:
Merge remote-tracking branch 'apache/master' into agenc-rowencoder

hvanhovell committed:
Fix tests

hvanhovell committed:
Add Set validation.

hvanhovell committed:
WrappedArray should use properly typed array

hvanhovell committed:
Merge remote-tracking branch 'apache/master' into SPARK-41993

hvanhovell committed:
Document lenient behavior.

hvanhovell committed:
Try to fix 2.13

hvanhovell committed:
Merge remote-tracking branch 'apache/master' into SPARK-41993

hvanhovell committed:
Code compiles

hvanhovell committed:
Merge remote-tracking branch 'apache/master' into JavaBeanEnc

# Conflicts:
#	sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala
#	sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala
#	sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/AgnosticEncoder.scala

hvanhovell committed:
Improve testing. Use Type instead of Class.

dongjoon-hyun reviewed (changes requested):
Could you check Scala 2.13 too? The original `AgnosticEncoders` PR broke Scala 2.13 with Kafka SQL module.
- https://github.com/apache/spark/pull/39517#pullrequestreview-1250710992

cloud-fan commented:
yea let's fix scala 2.13 first

hvanhovell committed:
Improve JavaTypeInference parameter resolution

# Conflicts:
#	sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/JavaTypeInferenceSuite.scala

hvanhovell committed:
Merge remote-tracking branch 'apache/master' into SPARK-42093

hvanhovell committed:
Fix type validation

hvanhovell committed:
Fix nullability

hvanhovell commented:
@cloud-fan @dongjoon-hyun can you take another look please?

cloud-fan commented:
thanks, merging to master/3.4!

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala:
```diff
@@ -166,317 +148,58 @@ object JavaTypeInference {
-            Invoke(inputObject, "name", ObjectType(classOf[String]), returnNullable = false))
-
-        case other =>
-          val properties = getJavaBeanReadableAndWritableProperties(other)
```
Hmm, I'm not sure if this has been discussed, but this is a breaking change causing customer issue like https://issues.apache.org/jira/browse/SPARK-48073.

viirya replied:
Is this an intentional change? @hvanhovell

viirya replied:
No matter what, this change has been there for a while. It seems difficult to change it back to old behavior. As discussed in the JIRA, I'd like to see if we can add a SQL conf to fall back to previous behavior of `Encoder.bean()`.

hvanhovell replied:
The behavior was apparently there before I got to it, this is PR that fixed it later: https://github.com/apache/spark/pull/42829

For the record there is no specification here, only tests.

viirya replied:
Thanks. After more internal discussion with the customer, we decided not to make a change to previous behavior with a config.
